exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/DrillSemiJoinRule.java

```diff
@@ -0,0 +1,182 @@
+    final RexBuilder rexBuilder = cluster.getRexBuilder();
+    if (project != null) {
+      final ImmutableBitSet bits =
+              RelOptUtil.InputFinder.bits(project.getProjects(), null);
+      final ImmutableBitSet rightBits =
+              ImmutableBitSet.range(left.getRowType().getFieldCount(),
+                      join.getRowType().getFieldCount());
+      if (bits.intersects(rightBits)) {
+        return;
+      }
+    }
+    final JoinInfo joinInfo = join.analyzeCondition();
+    if (!joinInfo.rightSet().equals(
+            ImmutableBitSet.range(aggregate.getGroupCount()))) {
+      // Rule requires that aggregate key to be the same as the join key.
+      // By the way, neither a super-set nor a sub-set would work.
+      return;
+    }
+    if (!joinInfo.isEqui()) {
+      return;
+    }
+    final RelBuilder relBuilder = call.builder();
+    relBuilder.push(left);
+    switch (join.getJoinType()) {
+      case INNER:
+        final List<Integer> newRightKeyBuilder = new ArrayList<>();
+        final List<Integer> aggregateKeys = aggregate.getGroupSet().asList();
+        for (int key : joinInfo.rightKeys) {
+          newRightKeyBuilder.add(aggregateKeys.get(key));
+        }
+        final ImmutableIntList newRightKeys = ImmutableIntList.copyOf(newRightKeyBuilder);
+        relBuilder.push(aggregate.getInput());
+        final RexNode newCondition =
+                RelOptUtil.createEquiJoinCondition(relBuilder.peek(2, 0),
+                        joinInfo.leftKeys, relBuilder.peek(2, 1), newRightKeys,
+                        rexBuilder);
+        relBuilder.semiJoin(newCondition);
+        break;
+
+      case LEFT:
+        // The right-hand side produces no more than 1 row (because of the
+        // Aggregate) and no fewer than 1 row (because of LEFT), and therefore
+        // we can eliminate the semi-join.
+        break;
+
+      default:
+        throw new AssertionError(join.getJoinType());
+    }
+    if (project != null) {
+      relBuilder.project(project.getProjects(), project.getRowType().getFieldNames());
+    }
+    call.transformTo(relBuilder.build());
+  }
+
+  /** DrillSemiJoinRule that matches a Join with an Aggregate (without agg functions) as its right
+   * child. */
+  public static class JoinToSemiJoinRule extends DrillSemiJoinRule {
+
+    /** Creates a JoinToSemiJoinRule. */
+    public JoinToSemiJoinRule(
+            Class<Join> joinClass, Class<Aggregate> aggregateClass,
+            RelBuilderFactory relBuilderFactory, String description) {
+      super(joinClass, aggregateClass, relBuilderFactory, description);
+    }
+  }
+
+  /**
+   * Check for the row schema if they aggregate's output rowtype is different
+   * from its input rowtype then do not convert the join to a semi join.
+   */
+  private static boolean isRowTypeSame(RelNode join, RelNode left, RelNode rightInput) {
+    return join.getRowType().getFieldCount() == left.getRowType().getFieldCount() +
```

Comment:
I am wondering if this should check for field count == 1 or leave it general ? If we leave it general, then should make sure to test correctness. My preference is to restrict it.